tests/Doctrine/Tests/ORM/Functional/Ticket/DDC2363Test.php

```diff
@@ -0,0 +1,303 @@
+        // orphanRemoval on Order::purchased_services_package and cascade="remove" on
+        // PurchasedServicesPackage::services.
+        $this->assertCount(
+            1,
+            $this->_em->getRepository(__NAMESPACE__ . '\\DDC2363ServicesPackage')->findAll()
+        );
+        $this->assertCount(
+            2,
+            $this->_em->getRepository(__NAMESPACE__ . '\\DDC2363Service')->findAll()
+        );
+
+        // Reinitializing EntityManager to simulate, for example, another HTTP request.
+        $this->_em->close();
+        $this->_em = $this->_getEntityManager();
+
+        // We load again the Order...
+        $order = $this->_em->getRepository(__NAMESPACE__ . '\\DDC2363Order')->find(1);
+
+        // To make test pass, uncomment the following line. So ServicesPackage will be hydrated.
+        // $order->getServicesPackage()->getName();
+
+        // ... and create another ServicesPackage with two Services.
+        $servicesPackage = $this->createNewServicesPackage($this->_em);
+        $order->setServicesPackage($servicesPackage);
+
+        $this->_em->persist($order);
+        $this->_em->flush();
+        $this->_em->flush(); // <--- This is the problem!
+
+        // There should be one ServicesPackage (and two Services) again, right?
+        $this->assertCount(
+            1,
+            $this->_em->getRepository(__NAMESPACE__ . '\\DDC2363ServicesPackage')->findAll()
+        );
+        $this->assertCount(
+            2,
+            $this->_em->getRepository(__NAMESPACE__ . '\\DDC2363Service')->findAll()
+        );
+    }
+
+    private function createNewServicesPackage()
+    {
+        $serviceA = new DDC2363Service();
+        $serviceA->setName('BASE SERVICE A');
+
+        $serviceB = new DDC2363Service();
+        $serviceB->setName('BASE SERVICE B');
+
+        $servicesPackage = new DDC2363ServicesPackage();
+        $servicesPackage->setName('SERVICES PACKAGE');
+        $servicesPackage->addService($serviceA);
+        $servicesPackage->addService($serviceB);
+
+        return $servicesPackage;
+    }
+
+}
+
+/** @Entity */
+class DDC2363Order
+{
+    /** @Id @Column(type="integer") @GeneratedValue **/
+    protected $id;
+
+    /** @Column(type="string") **/
+    protected $name;
+
+    /**
+     * @OneToOne(
+     *   targetEntity="DDC2363ServicesPackage",
+     *   inversedBy="order",
+     *   cascade={"persist", "remove"},
+     *   orphanRemoval=true
+     * )
+     */
+    protected $services_package;
+
+    public function setId($id)
+    {
+        $this->id = $id;
+    }
+
+    public function getId()
+    {
+        return $this->id;
+    }
+
+    public function setName($name)
+    {
+        $this->name = $name;
+    }
+
+    public function getName()
+    {
+        return $this->name;
+    }
+
+    public function setServicesPackage(DDC2363ServicesPackage $services_package)
+    {
+        $this->services_package = $services_package;
```

Comment:
In most cases, I'd set the order property on the existing record to null here to ensure that the object graph in memory is correct. This would sidestep the issue above as the lazy loading wouldn't be triggered in the cascade remove operation. It also means that any code with a reference to the old services package object can't use it (assuming you put some guards in place to throw exceptions if the order property is null)

Reply:
That's a workaround though, not a fix

Reply:
I'd argue that you are ensuring correct system behaviour. If you weren't persisting to the DB; you should set the order property null on the old object as it is no longer linked to one. Otherwise you may end up with weird edge cases if something else in your system maintains a reference to the old object. (Which is actually what we are seeing here from a different angle)

Reply:
Ah, I see what you mean, so the lazy-loading no longer occurs in ORM-internals but rather when the inverse side is set to `null` by the code here.

That could work, yes.

I'm wondering if the test scenario in this particular ticket has a valid use-case scenario then.

Also note that this will break if you have logic like following (valid, IMO):

```php
    public function setServicesPackage(DDC2363ServicesPackage $services_package)
    {
        $originalServicesPackage = $this->services_package;
        $this->services_package = $services_package;
        $services_package->order = $this;
        $originalServicesPackage->order = null; // lazy-loading will reset the property
    }
```

Reply:
Hi @Ocramius, I don't remember the whole story about this issue but I remember that we had this issue in production on a real (big) project so I think it's a real use case. Why do you wonder if it's a valid use-case? I mean, with correct orphan removal configuration I'd expect to have those orphan removed.

Reply:
It's a question of references to be cleaned. This code is wrong:

```php
$this->services_package = $services_package;
```

This code is right:

```php
if ($this->services_package) {
    $this->services_package->order = null;
}

$this->services_package = $services_package;
```

This doesn't deny the ORM bug, it just denies this specific scenario (the one in the test case).

Reply:
@Ocramius so even with the right code there is a bug?

Reply:
Yes, but I need a test scenario that doesn't have your userland bug in it.

In my opinion, the bug is that fetches are being performed during a `flush()` (that's already super-weird, but understandable)

Reply:
I'd suggest that the issue may lie in lazy loading overwriting a property which has been changed. Is it possible to fix that somehow?

Reply:
We could disable fetc-joining during flush operations, but I need to check if that's even configurable.

Reply:
@Ocramius so do you need that I push the right code on my fork for this PR?

Reply:
No need - the issue is clear, I provided a newer/rebased PR for it, but the action to be taken is still unclear.

Reply:
Ok.